storagenode/piecestore/endpoint.go

```diff
@@ -551,6 +553,50 @@ func (endpoint *Endpoint) SaveOrder(ctx context.Context, limit *pb.OrderLimit, o
 	}
 }
 
+// Retain keeps only piece ids specified in the request
+func (endpoint *Endpoint) Retain(ctx context.Context, retainReq *pb.RetainRequest) (*pb.RetainResponse, error) {
+	peer, err := identity.PeerIdentityFromContext(ctx)
+	if err != nil {
+		return nil, Error.Wrap(err)
+	}
+
+	filter, err := bloomfilter.NewFromBytes(retainReq.GetFilter())
+	if err != nil {
+		return nil, Error.Wrap(err)
+	}
+
+	limit := 20
+	offset := 0
+	piecesRemaining := true
+
+	for piecesRemaining {
+		pieceIDs, err := endpoint.pieceinfo.GetPiecesID(ctx, peer.ID, retainReq.GetCreationDate(), limit, offset)
+		if err != nil {
+			return nil, Error.Wrap(err)
+		}
+		for _, pieceID := range pieceIDs {
+			if !filter.Contains(pieceID) {
+				err = errs.Combine(err, endpoint.store.Delete(ctx, peer.ID, pieceID))
+				err = errs.Combine(err, endpoint.pieceinfo.Delete(ctx, peer.ID, pieceID))
+			}
+		}
+		piecesRemaining = (len(pieceIDs) == limit)
+		offset += limit
+		runtime.Gosched()
```

Comment:
Let's add a comment on why we call `runtime.Gosched()` so it is not removed by accident in the future. The reason is that the GC process is expected to be long, we want to keep it at low priority, so other goroutines can continue serving requests.